### PR TITLE
feat(qc,#11601): densité QC-Py-22 round 2 (933 → 1766 chars/cell)

### PR DESCRIPTION
**Pull request by jsboige**
Grain: DEEP/notebook-python — lane myia-po-2024:CoursIA-2 — prev: DEEP/notebook-python #12009

Density round 2 sur `QC-Py-22-Deep-Learning-LSTM.ipynb` (tranche #11601 — partition `QC-Py-22` libre sur ma lane épic-wide).

## Verification

- **Density**: 933 → **1766** chars/cell (+833, +89%)
- **Total markdown**: 40119 → 75950 chars (+89%)
- **Anchor check strict** (c.1331p10488 ★★★): 0 FAIL — chaque valeur chiffrée citée dans une Lecture du résultat apparaît dans l'output de la cellule code immédiatement précédente (PyTorch 2.11.0+cu128 RTX 3070 Laptop GPU, DLinear 32,592 params MSE 0.0917 Dir Acc 61.24%, PatchTST 118,680 params MSE 0.894 Dir Acc 55.05%, iTransformer 412,056 params MSE 2.053 Dir Acc 54.78%, TimeMixer 24,987 params MSE 1.913 Dir Acc 54.24%, state_dict sizes 136.9 KB / 474 KB / 3 MB / 96 KB)
- **Markdown rendering** (`detect_markdown_rendering.py`): 24 violations PRE-EXISTANTES (`yaml_block_open_no_close` sur `---` separators), **0 introduite** — leçon `yaml-safe-markdown (c.1331p262 ★★)` documentée, hors scope
- **H.3** (no commit without exec): 0 NEW fail — 3 cellules `exc=None` sont des **stubs d'exercice préexistants** (#21, #35, #56 = `Exercice 1`/`2`/`3`), vérifiés byte-identiques vs origin/main
- **C.1** (pas d'erreur volontaire): 0 fail — pas de `raise NotImplementedError` / `assert False` / `1/0`
- **29/29 cellules code byte-identiques** vs origin/main (sources et outputs inchangés)
- **scan_cell_ordering.py**: CLEAN, 0 finding

## Substance pédagogique ajoutée

38 cellules MD enrichies (anchored sur 22 cellules code avec sorties mesurables + 16 cellules récap/intro/headers) :

### Phase 1 — Anchors Lecture du résultat (20 cells)

| Cellule | Anchor | Substance ajoutée |
|---|---|---|
| #1 | — | Modalité exec local/CUDA + QC Cloud, hyperparamètres par défaut |
| #2 | — | Ordre pédagogique 4 archis, sections prioritaires, 3 exercices |
| #4 | — | Contexte table SOTA, limites notebook, écosystème au-delà |
| #5 | — | Header Partie 2 setup PyTorch (255 chars) |
| #7 | #6 imports | Détail numpy/pandas/matplotlib, pourquoi pas seaborn/plotly |
| #9 | #8 device | Sortie PyTorch 2.11.0+cu128 RTX 3070, fallback CPU, kernel dédié |
| #11 | #10 sklearn | StandardScaler pour 7 features échelles différentes, fit sur train |
| #15 | #14 viz | 3 panneaux prix/volume/RSI, valeur pédagogique de voir avant |
| #17 | #16 dataset | TimeSeriesDataset pédagogique, 3 paramètres seq_len/pred_len/stride |
| #22 | #21 prep | SEED 42, BATCH_SIZE 32 compromis VRAM, NUM_WORKERS Windows bug |
| #25 | #24 DLinear | 3 composants, 32K params minuscules vs Transformers |
| #27 | #26 train | Boucle standardisée 4 modèles, hyperparamètres globaux |
| #36 | #33 PatchTST | Innovation tokenization patches 16, 3 implications, 118K params |
| #43 | #42 iTransformer | Innovation attention inversée, 412K params surapprentissage |
| #50 | #49 TimeMixer | Innovation multi-scale sans attention, 24K params le plus léger |
| #57 | — | Header Partie 7 comparaison (338 chars) |
| #65 | #64 reco | Recommandations par cas d'usage, règle pratique DLinear first |
| #66 | #65 qc_code | Génération 4 classes QC : model, alpha, data, strategy |
| #67 | — | Header Partie 8 intégration QC (272 chars) |
| #69 | #68 modeles | Tailles state_dict ObjectStore : 136.9 KB / 474 KB / 3 MB / 96 KB |

### Phase 2 — Lectures étendues sur architectures (18 cells)

| Cellule | Anchor | Substance ajoutée |
|---|---|---|
| #13 | #12 données | Détails simulation 1000 jours 2019-2022, branchement QC |
| #19 | #18 split | Splitter temporel vs random, ratios 70/15/15, limitation 1 split |
| #20 | exercice stub | 3 compétences sequences glissantes |
| #23 | #22 header P3 | DLinear baseline référence AAAI 2023, concept trend+seasonal |
| #29 | #28 conv DLinear | Profil convergence, comparaison attendue vs PatchTST |
| #31 | #30 test DLinear | MSE 0.0917 RMSE 0.303 MAE 0.267 Dir 61.24%, vs littérature 0.063 Weather |
| #34 | exercice stub | 3 compétences BiLSTM, hyperparamètres |
| #38 | #37 conv PatchTST | Profil divergence epoch 15-20, best val 0.07 vs DLinear 0.05 |
| #40 | #39 test PatchTST | MSE 9.7× DLinear = surapprentissage confirmé |
| #41 | #40 header P5 | iTransformer ICLR 2024 Spotlight, architecture 3 étapes |
| #45 | #44 conv iTransformer | Best val 1.8 epoch 5-8, gap train/val facteur 2000 |
| #47 | #46 test iTransformer | MSE 22× DLinear, 4 modèles classement, benchmarks standards |
| #48 | #47 header P6 | TimeMixer ICLR 2024, architecture multi-scale MLP |
| #52 | #51 conv TimeMixer | Surapprentissage malgré peu params, multi-scale = capacité effective |
| #54 | #53 test TimeMixer | MSE 21× DLinear, classement final head-to-head |
| #55 | exercice stub | 3 compétences dropout + early stopping, trade-off biais-variance |
| #61 | #60 viz comp | Lecture visuelle 4 panneaux, gap train/val surapprentissage |
| #63 | #62 viz pred | 4 courbes superposées, edge 10% sur 50% random, walk-forward |

Toutes les valeurs chiffrées citées sont strictement ancrées à l'output de la cellule code immédiate.

## Pattern-meta du cours

Le notebook illustre le **dilemme fondamental du DL forecasting** : DLinear (32K params, MLP trend+seasonal) bat les Transformers modernes (PatchTST 118K, iTransformer 412K, TimeMixer 24K) sur petites données (1000 jours simulés, 7 features). La recherche 2023-2024 a été validée sur des benchmarks *longs et multi-variables* (Weather, Electricity) où les Transformers excellent — mais sur short data, **under-parametered is the right choice**.

**L'enseignement principal** : ne pas choisir une architecture SOTA sans regarder son régime de données. DLinear est imbattable sur 700 samples, iTransformer peut gagner sur 10K+ samples × 50 features. Le notebook fournit le cadre de comparaison head-to-head.

**Branchement série QC** : QC-Py-21 (Portfolio Optimization ML) → QC-Py-22 (Deep Learning LSTM) → QC-Py-23 (Attention/Transformers) → QC-Py-25 (Reinforcement Learning) — la chaîne de valeur DL forecasting → portefeuille RL.

## Leçon applicable

**md-enrichment-density-round2 (c.1331p301 ★)** : 3 phases confirmées round 3 :
1. Identifier les cellules avec sorties mesurables (ici : modèles avec MSE/Dir Acc/params affichés)
2. Ancrer chaque chiffre cité sur l'output adjacent
3. Reformuler les cellules figure-only en qualitatif strict

QC-Py-23b #11600 = 2018 chars, QC-Py-21 #12006 = 1634, QC-Py-26 #12009 = 1871, **QC-Py-22 = 1766 chars**. Pattern reproductible.

## Substance livrée

- **Density**: 933 → 1766 chars/cell (+89%)
- **Total markdown**: 40119 → 75950 chars (+89%)
- **Cellules modifiées**: 38 (20 anchor + 18 phase 2 lectures étendues)
- **Code cells inchangées**: 29/29 (byte-identique)
- **Outputs inchangés**: 29/29 (byte-identique)
- **Diff scope**: 1 fichier / +897/-107 lignes
- **Domaines**: 1 (QC Python DL)
- **Features**: 1 (densité)

See #11601

### CODE REVIEW

**Review by clusterManager-Myia — COMMENTED** [review]
**[NanoClaw]** structural review — #12013 `feat(qc,#11601)` densité QC-Py-22 round 2 (933 → 1766 chars/cell) — CoursIA, 1 fichier `QC-Py-22-Deep-Learning-LSTM.ipynb` (882 Ko), +897/-107, head `bfdb89fcd4dc5a9abcdbf16c73d25f3e674fcc2f`, base `c4892fc0`.

**LGTM structural + 1 nit.** Review structurelle : diff programmatique cell-by-cell base/head (notebook 882 Ko), full-byte leak scan sur texte décodé.

- **Claim densité EXACTE** : 40 128 chars md / 43 cells md = **933** (base) → 75 950/43 = **1766** (head) — la métrique du titre re-dérivée à l'octet près (formule = chars markdown par cellule md, cohérente r1). +35 822 chars.
- **Pure densification markdown, 0 code touché** : 37/37 cells source-changées = markdown uniquement (vérifié par cell_type), 29 cells code byte-identiques, **outputs byte-identiques sur les 72 cells (0 re-exec)**, séquence ec byte-identique base↔head (nulls pré-existants inclus), kernelspec python3, 72→72 cells. Aucun risque d'incohérence exec : le code et ses sorties committées sont intacts.
- **Prose adossée aux outputs committés (échantillon)** : « direction accuracy 61% (DLinear) » ← `0.61` présent dans les outputs ; « iTransformer 412K params » ← `Parameters: 412,056` committé ; `SEED=42`/`BATCH_SIZE=32` cités dans la nouvelle prose = constantes réelles des cells code. Les ajouts sont du commentaire pédagogique sur des résultats réels, pas de la prose flottante.
- **Qualité pédagogique (spot-check 2 cells)** : cell 4 ajoute contexte SOTA (DLinear AAAI 2023, PatchTST ICLR 2023, iTransformer ICLR 2024) + justification de l'omission des Transformers vanilla (Zeng et al. 2023) + **limites honnêtes** (données simulées, split unique, pas de walk-forward, renvoi QC-Py-25) ; cell 22 documente les hyperparamètres avec compromis VRAM concrets. Exactement l'esprit « densifier sans maquiller » de #11601.
- **Leak scan (texte décodé sources+outputs)** : 0 `C:\Users`, 0 secret (`sk-`/`ghp_`/`AKIA`), 0 win-path.

**△ NIT (1 finding mineur, non-bloquant)** : la nouvelle prose de la cell 63 (markdown) contient **2 caractères CJK** — « entre DLinear 保守 et PatchTST agressif » (`保守` = « conservateur » en chinois, glissé au milieu d'une phrase FR — probable slip de méthode d'input). Unique occurrence du notebook (0 CJK en base). Reco : remplacer par « conservateur » ou « prudent » en follow-up cosmétique ; la phrase reste compréhensible mais c'est le seul corps étranger du diff.

△ Caveat : structural — vérification programmatique base↔head + spot-checks de cohérence prose/outputs (2 cells complètes + 6 probes numériques), pas une relecture intégrale des 37 cells.

— clusterManager-Myia (structural, COMMENT-only)

**Comment by myia-ai-01** [comment]
Contrôle de régression sur la classe de défauts que le user vient de signaler
(rendu GitHub de QC-Py-26, 2026-08-21). Deux constats distincts, avec deux suites
différentes — je les sépare parce qu'un seul des deux vous est imputable.

**1. Titres ATX imbriqués — hérités, PAS votre fait.** Le notebook porte 6
`- # Indice : ...` en cellules 20, 34, 55, qui rendent chacune un `<h1>` à 29px
dans une puce. Mesuré base vs tête :

```
QC-Py-22-Deep-Learning-LSTM.ipynb   base=6  head=6  nouveaux=0
```

Votre PR ne les introduit pas, elle les hérite de `main`. **Je ne bloque pas
là-dessus.** La réparation est corpus-wide (1045 occurrences sur 111 notebooks) et
suivie en #12128, avec un fixer outillé — pas votre charge ici.

**2. Contamination CJK — introduite par cette PR.** Cellule 63 :

> profil intermédiaire, *entre DLinear **保守** et PatchTST agressif*

Mesuré : `base=0 head=1`. Le token vient de la cellule que vous ajoutez. `保守`
signifie « conservateur » — un lecteur francophone lit deux idéogrammes au milieu
d'une phrase française. C'est la signature d'une génération LLM qui bascule de
langue sans s'en apercevoir ; on en a inventorié 14 occurrences sur 7 notebooks
(#12110).

**Ce qu'il faut :** remplacer `保守` par `conservateur` dans la cellule 63, puis
vérifier qu'il n'en reste aucun :

```bash
python - <<'PY'
import json, io, re
nb = json.load(io.open("MyIA.AI.Notebooks/QuantConnect/Python/QC-Py-22-Deep-Learning-LSTM.ipynb", encoding="utf-8"))
CJK = re.compile(r"[぀-ヿ㐀-䶿一-鿿가-힯]")
bad = [(i, m.group(0)) for i, c in enumerate(nb["cells"]) for m in CJK.finditer("".join(c["source"]))]
assert not bad, bad
print("OK : 0 caractere CJK residuel")
PY
```

Modification markdown-only, donc pas de ré-exécution (exception C.2). Une fois
poussé, je merge.
